Posix x86_64 Calling conventions fixes

SSoulaimane opened the pull request:
Make DMD adhere better to the platform native calling conventions on x86_64 posix targets

SSoulaimane committed:
Extend genmovreg() to support exchange between SSE and general purpose registers

SSoulaimane committed:
Make use of @kinke classification pr (#8837)

SSoulaimane committed:
Simplify register exchange by using genmovreg()

SSoulaimane committed:
Base the register selection on the type of the first part of a struct
This what cgelem.d:elstruct() does

SSoulaimane committed:
Introduce A separate function for allocating return registers

SSoulaimane committed:
Use the proper registers for function returns
Allocate the proper registers in both sides of a function call
the caller and the calle.
Also generate some register exchange code as a form of adapter for the current codgen API. An example would
cfloat, The current codgen API expects cfloat to be in two registers whereas in 64 bits posix it is returned in only one register.

SSoulaimane committed:
Enable fpu counter assertion with an exception
allow a valid case where stackused == 1
the fpu stack may be preloaded in case the function call
is the second operand to some binary operator
example: `5 + fun()`
ST(0) would be holding the constant 5 prior to executing funccall()

SSoulaimane committed:
Special case of a struct > 16 bytes when a it only contains one field which is a SIMD vector

SSoulaimane committed:
SIMD 32 bytes variadic arguments should be passed on the stack
According to the System V x86_64 ABI spec

SSoulaimane committed:
For complex float80 make sure ST(0) and ST(1) hold the real and imaginary part accordingly
This is only required on Posix x86_64

SSoulaimane committed:
Special handling for complex float32 packed in a single register
The codgen API expect complex numbers to be in two separate registers

SSoulaimane committed:
Treat static arrays the same as structs.

SSoulaimane committed:
A better fix for past issues with casting a SIMD vector to a static array.
Array operations will require moving the SIMD vector out of the SIMD register first.

SSoulaimane committed:
x87 cases not forwarded correctly
runnable/complex.d:test7581() breaks on assert(stackused == 0);
This caused return values int ST0 not to be popped off the fpu stack
when only the flags are needed (ex: comparison to zero).

SSoulaimane commented:
All test cases not related to this PR pass. Two tests fail `testargtypes.d` and `testabi.d` which I'm working on now. I did hand checks though.  This isn't ready for merge yet but it's ready for a review.

thewilsonator commented on README.md:
```diff
@@ -1,29 +1,8 @@
-"I hereby assign copyright in my contributions to DMD to the D Language Foundation"
-
-and include your name and date.
+This branch includes fixes to DMD's calling conventions on platforms that follow the System V 64 bit ABI.
```
Please restore this file.

SSoulaimane replied:
sure

SSoulaimane replied:
done

thewilsonator commented on src/dmd/backend/ty.d:
```diff
@@ -183,6 +183,7 @@ enum
     mTYshared       = 0x00100000,    // shared data
     mTYnothrow      = 0x00200000,    // nothrow function
 
+    mTYreplaced     = 0x01000000,    // scalar replacement for aggregate
```
Is the collision with the below flags intentional?

SSoulaimane replied:
Yes. Although is says in the comment "Used only by C/C++ compiler" I still can't find any of them in the backend. maybe there were used in some previous version of the backend.

thewilsonator replied:
Hmm. @WalterBright ?

ibuclaw replied:
Likely to be in the dmc repo if anywhere.

thewilsonator commented on src/dmd/e2ir.d:
```diff
@@ -188,6 +188,7 @@ private elem *callfunc(const ref Loc loc,
         /* Convert arguments[] to elems[] in left-to-right order
          */
         const n = arguments.dim;
+        size_t nparams = Parameter.dim(tf.parameters);
```
This code motion seems unnecessary

SSoulaimane replied:
It isn't necessary but I felt it didn't belong inside the loop either so I moved it out.

thewilsonator replied:
Fair enough, Walter might have different opinions though.

thewilsonator commented on src/dmd/globals.d:
```diff
@@ -242,6 +242,25 @@ struct Param
 alias structalign_t = uint;
 
+/// target follows System V 64bit ABI
+bool isSysV64bitABI(scope ref Param params)
```
`const`

SSoulaimane replied:
sure

thewilsonator commented:
Thanks! This will need to be squashed significantly before merge, but for now if you could squash the typo and minor mixup commits that would be great.

jacob-carlborg commented:
What exactly is this doing? Changing DMD to **not** use its own calling conventions?

thewilsonator commented:
this is for `extern(C[++])`, the current posix x64 ABI is broken, this fixes it.

ibuclaw commented on src/dmd/globals.d:
```diff
@@ -242,6 +242,25 @@ struct Param
 alias structalign_t = uint;
 
+/// target follows System V 64bit ABI
+bool isSysV64bitABI(scope ref const Param params)
```
Wrong place to put this.

SSoulaimane replied:
True.

ibuclaw commented on src/dmd/target.d:
```diff
@@ -519,9 +520,14 @@ struct Target
      */
     extern (C++) static TypeTuple toArgTypes(Type t)
     {
+        if (global.params.isSysV64bitABI)
```
Stop putting target-specific information in `global.params` please.

ibuclaw replied:
As a side note, this also has a side effect, but you don't have any parentheses present to denote this.

SSoulaimane replied:
I see. Thank you.

thewilsonator commented:
This will also need to enumerate the bugs that it fixes in one of the commit messages in order to close them, take a look at some the other bug fix PRs for that.

n8sh commented on src/dmd/argtypes.d:
```diff
@@ -20,6 +20,11 @@ import dmd.globals;
 
+private bool isDMDx64Target()
+{
+    return global.params.is64bit;
```
For what purpose was this function created?

jacob-carlborg commented on src/dmd/argtypes_sysv_x64.d:
```diff
@@ -0,0 +1,394 @@
+ *      A tuple of zero length means the type cannot be passed/returned in registers.
+ *      null indicates a `void`.
+ */
+extern (C++) TypeTuple toArgTypes_sysv_x64(Type t)
```
The name does not follow the D style/naming conventions.

jacob-carlborg commented on src/dmd/argtypes_sysv_x64.d:
```diff
@@ -0,0 +1,394 @@
+    if (c0 == Class.X87)
+        return new TypeTuple(Type.tfloat80);
+    if (c0 == Class.ComplexX87)
+        return new TypeTuple(Type.tfloat80, Type.tfloat80);
```
Perhaps using a switch for these three if statements instead.

jacob-carlborg commented on src/dmd/argtypes_sysv_x64.d:
```diff
@@ -0,0 +1,394 @@
+    X87Up,
+    ComplexX87,
+    NoClass,
+    Memory
```
Enum member names should start with a small letter.

jacob-carlborg commented on src/dmd/argtypes_sysv_x64.d:
```diff
@@ -0,0 +1,394 @@
+    if (a == Class.X87 || b == Class.X87 ||
+        a == Class.X87Up || b == Class.X87Up ||
+        a == Class.ComplexX87 || b == Class.ComplexX87)
+        return Class.Memory;
```
Perhaps a switch instead.

jacob-carlborg commented on src/dmd/argtypes_sysv_x64.d:
```diff
@@ -0,0 +1,394 @@
+
+    this(size_t size)
+    {
+        assert(size > 0);
```
Perhaps this should be in a contract instead, i.e. an `in` block.

thewilsonator replied:
I don't think contract are used in dmd.

jacob-carlborg replied:
Is there a reason we can't start 😃.

jacob-carlborg commented:
All new public symbols should have Ddoc comments.

SSoulaimane commented:
Individual commit do make sense now after some reordering and squashing. The more significant commits are the older ones while the recent ones are mostly small fixes (that still make sense).

WalterBright commented:
This is extremely difficult to review because it does way too much. It should be done incementally in smaller steps.

thewilsonator commented:
> This is extremely difficult to review because it does way too much.

@WalterBright This is going to get reviewed by me irrespectively. I would recommend review-by-commit here.

> It should be done incementally in smaller steps.

 @SSoulaimane While nice in practice, focus on getting all the tests to pass first. I'll deal with the fission of this for merge and review when the time comes. The restructuring you did of those commits is a good start. Thanks again for fixing these nasty bugs!

kinke commented:
> Two tests fail `testargtypes.d` and `testabi.d`

Fixing `testargtypes.d` is trivial, just get rid of all `version (DigitalMars)` blocks. These are cases where the old `toArgTypes()` for 64-bit targets diverges from the new `toArgTypes_sysv_x64()` one.

SSoulaimane committed:
Testsuite: Get it to compile
DMD complains about global variable access from asm code

SSoulaimane committed:
Testsuite: Mixed cases incorrectly classified as Integer cases

SSoulaimane committed:
Testsuite: Padding maybe non-zero

SSoulaimane committed:
Testsuite: Include tests for float80 based types

SSoulaimane committed:
Testsuite: Include SIMD tests

SSoulaimane committed:
Testsuite: quit early avoid some errors

SSoulaimane committed:
Testsuite: assert on test failures

SSoulaimane committed:
Testsuite: update debug prints

SSoulaimane committed:
Remove obsolete argtypes tests

thewilsonator commented:
In general prefer rebasing, not that it really matters here, the history will get severely altered before merge anyway. Keep up the great work!

SSoulaimane commented:
> In general prefer rebasing, not that it really matters here, the history will get severely altered before merge anyway. Keep up the great work!

Right. I'll do another rebase squash round once I get everything up and running.

SSoulaimane committed:
Temporarily Revert "Swap real and imaginary for float80 for 64 bits targets"
Just for CI's sake

This reverts commit b4297b2f6a5395f4839b52d690e726c1ab3774b3.

SSoulaimane commented:
Thanks @thewilsonator I was wondering about these failures myself I'm not sure yet about the last two phobos failures though the second one the std/uni.d seems to come and go on Posix x64 systems depending on the OS configuration. I'll investigate further.
The semaphore failure doesn't look scary so I'm delaying it right now it until I figure out those more obscure ones.

SSoulaimane committed:
In LP32, slices and delegates have size 8 and alignment 4

SSoulaimane committed:
Type preservation with inlining
When an if statement gets inlined into a comma expression and none of the if or else branches have a return statement then one of them may get removed by the optimizer which causes a type loss on the else branch.
Q: Why does this matter?
A: If the type is lost then the wrong registers will be allocated

thewilsonator commented:
All green, nice!

SSoulaimane commented:
Yes! At last. I think I should take out some generic bug fix commits out as separate pull requests. The next step I guess is heading to bugzilla.

thewilsonator commented:
> I think I should take out some generic bug fix commits out as separate pull requests.

That would be a good start, any refactoring one as separate PR would also make things easier.

SSoulaimane commented:
Sure. Otherwise If Anything I can do or If you have any questions or comments just go a head. This PR does indeed as @WalterBright said touch on many code yet I tried to keep changes at a minimum. In fact I didn't even take the more straightforward approach to solve this problem which is to make the current SRA facilities more heterogeneous aware I didn't do that because it would require a lot more changes. In reality I spent most of the time dealing with side issues which used to be working before.

SSoulaimane commented on src/dmd/backend/cgxmm.d:
```diff
@@ -1573,6 +1576,11 @@ bool xmmIsAligned(elem *e)
+    else if (e.Eoper == OPind)
+    {
+        // can't know for sure until runtime
+        return false;   // assume not aligned
```
First we can't really assume random pointers to be aligned to any boundary.

But to illustrate why we should be pessimistic here consider the following program:
```D
import core.simd;
void main()
{
    struct S { long4 y; }
    S v; v.y += 1; // crashes
}
```
if you compile this little program with the current DMD and run it, it will crash obscurely, usually it will say something like `Illegal instruction`. If you run the program on [Intel's emulator](https://software.intel.com/en-us/articles/intel-software-development-emulator/) you'll get a more useful error message. For example on my machine it says:
```
SDE ERROR:  TID: 0 executed instruction with an unaligned memory reference to address 0x7ffffddfc8b0 INSTR: 0x55ce337f463a: IFORM: VMOVDQA_MEMqq_YMMqq :: vmovdqa ymmword ptr [rbp-0x20], ymm1
	IMAGE:    /home/user/simd
	FUNCTION: _Dmain
	FUNCTION ADDR: 0x55ce337f460c
```
It complains about unaligned memory. However the address `rbp-32` which is is this case `0x7ffffddfc8b0` is in fact aligned to `16` bytes (`0x7ffffddfc8b0 % 16 == 0`) because it is required for the stack to be aligned to `16` bytes upon function entry (on Linux 64) but not aligned required to be aligned to `32` bytes thus our stack pointer isn't aligned to `32` (`0x7ffffddfc8b0 % 32 == 16`).

The CPU does not seem to tolerate misaligned pointers with instructions which expect aligned pointer therefor it is safer to use the slower instructions that work with both aligned and misaligned pointers.

SSoulaimane replied:
As for stack variable, perhaps we can force a 32 bytes stack alignment, this is what LDC does. LDC generates an instruction in the prolog of the function which looks like this:
```asm
and    rsp,0xffffffffffffffe0     # 0xffffffffffffffe0 == ~(32 - 1)
```

SSoulaimane replied:
I moved this into a separate pull request #9143

SSoulaimane commented on src/dmd/backend/cod1.d:
```diff
@@ -1836,10 +1836,10 @@ void fixresult(ref CodeBuilder cdb, elem *e, regm_t retregs, regm_t *pretregs)
                 reg = findreg(retregs & XMMREGS);
                 // MOVSD floatreg, XMM?
-                cdb.genxmmreg(xmmstore(tym), reg, 0, tym);
+                cdb.genxmmreg(xmmstore(tym, false), reg, 0, tym);
```
That is indeed true.

ibuclaw commented on test/runnable/testargtypes.d:
```diff
@@ -30,52 +30,59 @@ void chkPair(T,U,V)()
         chkIdentity!int();
-        chkIdentity!uint();
         chkIdentity!long();
-        chkIdentity!ulong();
```
Existing tests shouldn't be removed without good reason.

ibuclaw commented on test/runnable/testargtypes.d:
```diff
@@ -30,52 +30,59 @@ void chkPair(T,U,V)()
-version (X86_64)
+version (Posix) version (X86_64) version = Posix_X86_64;
+
+version (Posix_X86_64)
```
Wouldn't this mean that windows x86_64 abi is no longer being tested?

kinke replied:
These tests incl. Win64 were in vain/useless/confusing. They tested the SysV argtypes stuff on Windows, which is as wrong as it gets. But this should be discussed in #8837; I don't know why that commit gets a lot more feedback here than in its actual smaller PR.

TurkeyMan commented:
WIll this fix: https://issues.dlang.org/show_bug.cgi?id=19563 ?
Or: https://issues.dlang.org/show_bug.cgi?id=19179 ?

SSoulaimane committed:
use reg_t as type instead of int

SSoulaimane committed:
New strategy: Introduce a new pair of flags specifically for SROA with mixed registers
Use two new type flags for SROA `mTYxmmgpr` and `mTYgprxmm`
Since I figured I don't really need the full type of the original aggregate
preserved through optimization, but knowing which combination of registers to use is enough.

SSoulaimane committed:
Testsuite: void init

SSoulaimane committed:
Keep original type unchanged in setArray()
This was causing an ice in the backend

SSoulaimane committed:
Keep the old logic for Win64
Other parts (mainly isReturnOnStack) don't yet match the windows 64 ABI.

SSoulaimane committed:
Return in registers structs of size 3, 5, 7...

SSoulaimane committed:
keep backend aggregate type

SSoulaimane committed:
fix issue 20052

SSoulaimane committed:
fix issues 5570, 13957

SSoulaimane commented:
Rebooted at #10200.

12345swordy commented:
Why is this pr still open then?
